File: src/com/philhanna/flashcards/Session.java
package com.philhanna.flashcards;

import java.util.List;

import com.philhanna.flashcards.session.SessionCard;
import com.philhanna.flashcards.session.SessionCursor;

/**
 * One pass through the deck. Has methods that keep track of the current
 * card and the progress through the deck.
 * @see SessionCard
 * @see SessionCursor
 */
public interface Session {

   /**
    * Returns the current card showing
    * @return a session card object
    */
   public SessionCard getCurrentCard();

   /**
    * Returns the cursor object for this session, which lets us know how
    * far we have gone in the deck
    * @return the session cursor
    */
   public SessionCursor getCursor();

   /**
    * Moves forward in the deck to the next card
    */
   public void nextCard();

   /**
    * Moves backward in the deck to the previous card
    */
   public void previousCard();

   /**
    * Shuffles the deck
    */
   public void shuffle();

   /**
    * Reinserts a missed card in the deck
    */
   public void rotate();

   /**
    * Returns the cards that have been viewed in the current session
    */
   public List<SessionCard> getViewedCards();
   
   /**
    * Returns the number of seconds that have elapsed since the start of the session
    */
   public double getElapsedTime();
   
   /**
    * Returns the number of times cards have been presented for viewing
    */
   public int getCardViewCount();
}

File: src/flashcards/domain/session/session.py
# flashcards.domain.session.session

from abc import ABC, abstractmethod
from typing import Sequence

from flashcards.domain.session.session_card import SessionCard
from flashcards.domain.session.session_cursor import SessionCursor


class Session(ABC):

    @property
    @abstractmethod
    def current_card(self) -> SessionCard | None: ...

    @property
    @abstractmethod
    def cursor(self) -> SessionCursor: ...

    @abstractmethod
    def next_card(self) -> None: ...

    @abstractmethod
    def previous_card(self) -> None: ...

    @abstractmethod
    def shuffle(self) -> None: ...

    @abstractmethod
    def rotate(self) -> None: ...

    @property
    @abstractmethod
    def viewed_cards(self) -> Sequence[SessionCard]: ...

    @property
    @abstractmethod
    def elapsed_time(self) -> float: ...

    @property
    @abstractmethod
    def card_view_count(self) -> int: ...

    @abstractmethod
    def record_card_view(self) -> None: ...
